blas/blas64/blas64.go

```diff
@@ -98,64 +99,79 @@ type SymmetricPacked struct {
 
 // Level 1
 
-const negInc = "blas64: negative vector increment"
+const (
+	negInc    = "blas64: negative vector increment"
+	badLength = "blas64: vector length mismatch"
+)
 
 // Dot computes the dot product of the two vectors:
 //  \sum_i x[i]*y[i].
-func Dot(n int, x, y Vector) float64 {
-	return blas64.Ddot(n, x.Data, x.Inc, y.Data, y.Inc)
+func Dot(x, y Vector) float64 {
+	if x.N != y.N {
+		panic(badLength)
+	}
+	return blas64.Ddot(x.N, x.Data, x.Inc, y.Data, y.Inc)
 }
 
 // Nrm2 computes the Euclidean norm of the vector x:
 //  sqrt(\sum_i x[i]*x[i]).
 //
 // Nrm2 will panic if the vector increment is negative.
-func Nrm2(n int, x Vector) float64 {
+func Nrm2(x Vector) float64 {
 	if x.Inc < 0 {
 		panic(negInc)
 	}
-	return blas64.Dnrm2(n, x.Data, x.Inc)
+	return blas64.Dnrm2(x.N, x.Data, x.Inc)
 }
 
 // Asum computes the sum of the absolute values of the elements of x:
 //  \sum_i |x[i]|.
 //
 // Asum will panic if the vector increment is negative.
-func Asum(n int, x Vector) float64 {
+func Asum(x Vector) float64 {
 	if x.Inc < 0 {
 		panic(negInc)
 	}
-	return blas64.Dasum(n, x.Data, x.Inc)
+	return blas64.Dasum(x.N, x.Data, x.Inc)
 }
 
 // Iamax returns the index of an element of x with the largest absolute value.
 // If there are multiple such indices the earliest is returned.
 // Iamax returns -1 if n == 0.
 //
 // Iamax will panic if the vector increment is negative.
-func Iamax(n int, x Vector) int {
+func Iamax(x Vector) int {
 	if x.Inc < 0 {
 		panic(negInc)
 	}
-	return blas64.Idamax(n, x.Data, x.Inc)
+	return blas64.Idamax(x.N, x.Data, x.Inc)
 }
 
 // Swap exchanges the elements of the two vectors:
 //  x[i], y[i] = y[i], x[i] for all i.
-func Swap(n int, x, y Vector) {
-	blas64.Dswap(n, x.Data, x.Inc, y.Data, y.Inc)
+func Swap(x, y Vector) {
+	if x.N != y.N {
+		panic(badLength)
+	}
+	blas64.Dswap(x.N, x.Data, x.Inc, y.Data, y.Inc)
 }
 
 // Copy copies the elements of x into the elements of y:
 //  y[i] = x[i] for all i.
-func Copy(n int, x, y Vector) {
-	blas64.Dcopy(n, x.Data, x.Inc, y.Data, y.Inc)
+func Copy(x, y Vector) {
+	if x.N != y.N {
```

Comment:
I forgot to comment here. Since BLAS doesn't hold the notion of independent lengths here. The behaviour for `copy([n]T, [m]T)` is both unspecified and not obviously defined from external context. For example the built in `copy` does min elements, while here we panic. For the other diadic functions here there is a reasonable obvious choice, so I think they are fine. This one needs discussion and more documentation.

Reply:
I believe blas has an implied equality. It is always used as such in blas and lapack as far as I can remember.

Reply:
OK. That needs to be documented, at least on `Copy` since there is a semantic collision/overloading with `copy`.

Reply:
Done.